cql3/restrictions/statement_restrictions.cc

```diff
@@ -1286,16 +1286,26 @@ bool statement_restrictions::need_filtering() const {
         // clustering restrictions.  Therefore, a continuous clustering range is guaranteed.
         return false;
     }
-    if (!_clustering_columns_restrictions->needs_filtering(*_schema)) { // Guaranteed continuous clustering range.
-        return false;
-    }
-    // Now we know there are some clustering-column restrictions that are out-of-order or not EQ.  A naive base-table
-    // query must be filtered.  What about an index-table query?  That can only avoid filtering if there is exactly one
-    // EQ supported by an index.
-    return !(_clustering_columns_restrictions->size() == 1 && _has_queriable_ck_index);
 
-    // TODO: it is also possible to avoid filtering here if a non-empty CK prefix is specified and token_known, plus
-    // there's exactly one out-of-order-but-index-supported clustering-column restriction.
+    if (_has_queriable_ck_index && _uses_secondary_indexing) {
+        // In cases where we use an index, clustering column restrictions might cause the need for filtering.
+        // TODO: This is overly conservative, there are some cases when this returns true but filtering
```

Comment:
What is the **consequence** of this statement about "some cases when this returns true but filtering is not needed"? Are there cases where we shouldn't need ALLOW FILTERING but we incorrectly require it? If so, please add a xfailing test for that. If not, then please explain the consequence - just do-nothing filtering that slows down the query by a tiny bit?

Reply:
I saw now that below you do add a test for this case, but I suspect the test will fail on Cassandra, not on Scylla. I added a comment below why I think it should be reversed - the test should pass on Cassandra, but fail on Scylla until we fix Scylla's behavior (or, we can decide that we don't want to fix Scylla's behavior, but we'll need to make a conscious decision why).

Reply:
> What is the **consequence** of this statement about "some cases when this returns true but filtering is not needed"?

That depends entirely on the caller of this function.  The function itself shouldn't know the answer to that; the most it can say here is "in some cases I'll conservatively tell you you will need to filter, but actually the filter will not discard any of the rows you fetched".

> Are there cases where we shouldn't need ALLOW FILTERING but we incorrectly require it?

That depends on the spec for ALLOW FILTERING, which we know to be different from the need_filtering spec.  It should be possible in the future to check the need for ALLOW FILTERING without calling need_filtering().  The priority here should be fixing incorrect results, not performance warnings.